### ISSUE TITLE
Validation config errors when enabling the validation webhook

### ISSUE THREAD

**Issue by Stono** [istio/istio#28372]
**Bug description**
Hi, 
We're attempting to enable the validation config webhook.  This should have been an non-event for us as we've been using `istioctl validate -f` in our pipelines for months.  

However once enabled we were a alerted to a failure in `apiserver_admission_webhook_admission_duration_seconds_count`.

Looking into the `istiod` logs we see a single line: 
```
{"level":"info","time":"2020-10-28T17:26:10.966351Z","scope":"validationServer","msg":"configuration is invalid: gateway must have at least one server"}
```

But no context.

It seems to come from: https://github.com/istio/istio/blob/master/pkg/config/validation/validation.go#L321

**But we don't use gateways**, we have no gateway crds, we never have:
```
❯ kubectl get gateways.networking.istio.io --all-namespaces
No resources found
```

Therefore I've no idea what is failing? 

[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[x] Developer Infrastructure

**Expected behavior**
More context about the failure.

**Steps to reproduce the bug**
Unknown

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
1.16.12

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

**Comment by howardjohn** [comment]
How the webhook works is it doesn't enable itself (fail open) until it successfully blocks an invalid config. This was to avoid a constant struggle to reliable determine when to apply the webhook to avoid failures. It doesn't log on master though, I am not sure why it does here, will dig into it. But tl;dr is it should be a logging/monitoring bug

**Comment by Stono** [comment]
Is this a one time operation when `istiod` starts, or periodic? 

I ask as naturally we alert on rejected webhooks:

<img width="646" alt="Screenshot 2020-10-28 at 18 17 11" src="https://user-images.githubusercontent.com/1486729/97479142-d0200b00-1949-11eb-83cf-0920bf07eda4.png">

**Comment by howardjohn** [comment]
it should be once when starting the controller, which is generally once when istiod starts unless it loses leader election (could be temporary api server outage) which would restart it

**Comment by Stono** [comment]
If the purpose of this is to ensure that the controller gets an `Admission` event, you don't actually need to reject it, right?

If the controller receives your (obviously test) event, at that point you know the webhook is working - you could just not reject the bad config in that scenario and then switch to fail closed?

**Comment by howardjohn** [comment]
This is actually not fixed on master like I previously suggested

**Comment by howardjohn** [comment]
I am actually not sure how to do this. I wrote up what you suggested (in branch validation/no-fail) but then realized it will not work since the admission request can be handled by any Istiod instance. If we send a valid one, we cannot distinguish between "Created and went through webhook" or "created and did not go through webhook". The current mechanism works because we look at the error we get back from k8s

**Comment by howardjohn** [comment]
@jasonwzm any ideas?

**Comment by Stono** [comment]
Fundamentally if you're receiving **any** Admission events, the webhook is working?  so:

```
istiod 1&2 start; set webhook to failOpen
...
istiod1: created some valid crd implementation
istiod2: receives admission event, sets failOpen = false, subsequently deletes created test resource
```

**Comment by howardjohn** [comment]
Ah yeah the issue with the current implementation is only one is writing the webhook, but any instance may get the event. If any of them could update then it would work

**Comment by jasonwzm** [comment]
Is it a reproducible behavior? I could not find the logs in my GKE cluster. The implementation shows that we ignore the response from API server and only cares about if the config creation fails or not.

**Comment by howardjohn** [comment]
@jasonwzm yes, I can easily reproduce. I am not sure how you do not have the log. The api server calls the istiod validation server, and it logs all errors. It doesn't know it is a dry run. Also the metric Karl is looking at is from the api server, although istiod also emits its own metric that is also incremented

**Comment by Stono** [comment]
FYI this is still in `1.10.1` @howardjohn 

```
{"level":"info","time":"2021-06-21T07:46:44.247647Z","scope":"validationServer","msg":"configuration is invalid: gateway must have at least one server"}
```

**Comment by mohe2015** [comment]
Please reopen. This still seems to be an issue
```
2022-10-31T13:49:08.277174Z     info    validationServer        configuration is invalid: gateway must have at least one server
2022-10-31T13:49:08.279398Z     info    validationController    Endpoint successfully rejected invalid config. Switching to fail-close.
2022-10-31T13:49:08.285339Z     info    validationController    Successfully updated validatingwebhookconfiguration istiod-default-validator (failurePolicy=Fail,resourceVersion=16030)
2022-10-31T13:49:08.287951Z     info    validationController    validatingwebhookconfiguration istiod-default-validator (failurePolicy=Fail, resourceVersion=16030) is up-to-date. No change required.
2022-10-31T13:49:08.289869Z     info    validationController    validatingwebhookconfiguration istiod-default-validator (failurePolicy=Fail, resourceVersion=16030) is up-to-date. No change required.
```
Maybe at least adjust the error message so it's clear this is intentional.

**Comment by pre** [comment]
Came here by googling this error message occurring with istio 1.16.2 

`info	validationServer	configuration is invalid: gateway must have at least one server`

**Comment by rahul-paigavan** [comment]
@howardjohn I this issue is fixed through PR #44303 , but I am not able to find in which release it has be put into.
I'm using Istio 1.17 and seeing this issue! Can you please help me here!

**Comment by rahul-paigavan** [comment]
> @howardjohn I see this issue is fixed through PR #44303 , but I am not able to find in which release it has be put into. I'm using Istio 1.17 and seeing this issue! Can you please help me here!

Found the release, it is `1.18`.
